### DISCUSSION TITLE
Does the Leaf App Server expose a web API?

### DISCUSSION THREAD

**Discussion by artgoldberg** [uwrit/leaf#474] [Q&A] [answered]
We're engaged in a Security Risk Assessment for Leaf, so I need to describe its architecture and data flow.

The [Leaf architecture](https://leafdocs.rit.uw.edu/installation/) contains an App Server which performs all DBMS access and exposes a Leaf API.

Does the App Server also expose a web API? That is, can an authenticated http client issue http requests to the App Server and obtain DBMS query results?

Thanks
Arthur

**Comment by ndobb** [comment, marked as answer]
Right, in a typical 3-tier setup, the Leaf API runs on the App Server, and user HTTP requests are reverse-proxied to the API:

![image](https://user-images.githubusercontent.com/10424962/139310653-d0b16290-b9b4-41f6-8352-48b27a7f8f69.png)

Leaf uses a standard RESTful API, and most endpoints by and large are protected and only accept requests if a given HTTP request includes an `AccessToken` (example: [`/api/cohort/count`](https://github.com/uwrit/leaf/blob/master/src/server/API/Controllers/CohortCountController.cs#L21)). There is one and only one API.

`AccessTokens` are generated for a user only after they first are [authenticated and retrieve a `UserToken`](https://github.com/uwrit/leaf/blob/master/src/server/API/Controllers/UserController.cs#L64), after which they can [request an `AccessToken`](https://github.com/uwrit/leaf/blob/master/src/server/API/Controllers/UserController.cs#L80), which is valid for [only 6 minutes](https://github.com/uwrit/leaf/blob/master/src/server/API/Jwt/JwtProvider.cs#L119).

In other words, the flow is:
1. User requests a `UserToken` (i.e., authenticates), which is successful if and only if the request contains expected shibboleth headers with user name and groups.
2. User requests an `AccessToken` (valid for 6 minutes) using their `UserToken`.
3. User accesses sensitive (i.e., DBMS-using) API endpoints with the `AccessToken`.

This flow holds true whether calling from the Leaf web client (i.e., client -> Web Server -> App Server), directly via a shell within the App Server, or wherever. If the requester fails in either step (1) or step (2), the API will never allow them to do (3).

In other words, the API is hardened, and sensitive endpoints will **only trust requests that originate with an authenticated user and `AccessToken`, regardless of where the request comes from.**

For example, [`/api/config`]() does not require authentication, and I can call using curl from within the API server:
![image](https://user-images.githubusercontent.com/10424962/139313634-6650d9fd-f5b5-424c-a7f2-7c79081d9c77.png)

However I cannot do so with [`/api/cohort/<cohort_id>/demographics`](https://github.com/uwrit/leaf/blob/master/src/server/API/Controllers/CohortController.cs#L37), which requires a valid `AccessToken`:
![image](https://user-images.githubusercontent.com/10424962/139314744-3165f5b6-b7de-4482-b923-b3b31c68cd3a.png)

Please let me know if this makes sense or if you have other questions, Arthur.

Best,
-nic

**Reply by artgoldberg** [reply]
Thanks for the rapid, informative and extensive response, @ndobb . This is **definitely** an answer.

**Reply by artgoldberg** [reply]
However, the architecture diagram is a bit misleading. The Shibboleth module used by Apache doesn't communicate directly with the SAML Identity Provider. Instead, it [communicates through the client](https://en.wikipedia.org/wiki/Security_Assertion_Markup_Language#Use) (e.g. browser) via redirects to the SAML Identity Provider.

**Reply by ndobb** [reply]
That's true, and perhaps we can update to make that more clear. That diagram mostly exists to help new sites trying to make sense of Leaf at a high level grasp what it does, so it's not the best technical representation of auth procedures.
